File: A1/ex2.jl
using Pkg

Pkg.add("LightGraphs")
Pkg.add(name = "GraphIO", version = "0.5.0")
Pkg.add("PrettyTables")
Pkg.add("CSV")
Pkg.add("DataFrames")

using LightGraphs
using GraphIO
using PrettyTables
using Statistics
using CSV
using DataFrames

airports = Dict("PAR" => 2269, "LON" => 1713, "FRA" => 926, "AMS" => 116, "MOW" => 1954, "NYC" => 2161, "ATL" => 174, "BCN" => 240, "WAW" => 3234, "CHC" => 527, "DJE" => 726, "ADA" => 34, "AGU" => 63, "TBO" => 2877, "ZVA" => 3614)
nodes = collect(values(airports))
header = (["Airport", "Degree", "Strength", "Clustering coefficient", "Avgerage path length", "Maximum path length", "Betweeness", "Eigenvector centrality", "PageRank"])
table = []

g = loadgraph(joinpath("A1-networks/real/airports_UW.net"), NETFormat())

f(x) = gdistances(g, x)
degrees = degree(g, nodes)
# TODO: Add strength
strength = Vector{Float64}(undef, length(nodes))
cluster_coef = round.(local_clustering_coefficient(g, nodes), digits=8)
path_lengths = reduce(hcat, f.(nodes))'
average_path_lengths = round.(mean(path_lengths, dims = 2), digits=8)
max_path_lengths = maximum(path_lengths, dims = 2)
betweeness = round.(betweenness_centrality(g)[nodes], digits=8)
eigenvector = round.(eigenvector_centrality(g)[nodes], digits=8)
page_rank = round.(pagerank(g)[nodes], digits=8)

tab = hcat(collect(keys(airports)), degrees, strength, cluster_coef, average_path_lengths, max_path_lengths, betweeness, eigenvector, page_rank)

open("results/airports_descriptors_latex.txt", "w") do f
    pretty_table(f, tab; backend = Val(:latex), header = header)
end

CSV.write("results/airports_descriptors.csv", DataFrame(tab, :auto), header = header);

File: A1/igraph/ex2.R
library(igraph)
library(tibble)
library(dplyr)
library(xtable)

airports <-
  c(
    "PAR",
    "LON",
    "FRA",
    "AMS",
    "MOW",
    "NYC",
    "ATL",
    "BCN",
    "WAW",
    "CHC",
    "DJE",
    "ADA",
    "AGU",
    "TBO",
    "ZVA"
  )
g <-
  read_graph(file.path("..", "A1-networks", "real", "airports_UW.net"), format = "pajek")
g2 <- delete_edge_attr(g, "weight")

d <- degree(g2, V(g2)[airports])
s <- strength(g, V(g)[airports])
cc <- transitivity(g2, type = "local", vids = V(g2)[airports])
dists <- distances(g2, V(g2)[airports])
apl <-  rowMeans(dists)
mpl <-  apply(dists, 1, max)
e <- eigen_centrality(g2)$vector[airports]
b <- betweenness(g2, V(g2)[airports], normalized = TRUE)
pr <- page.rank(g2, vids = V(g2)[airports], directed = FALSE)$vector

t <-
  data.frame(
    "Degree" = d,
    "Stength" = s,
    "Clustering coefficient" = cc,
    "Average path length" = apl,
    "Maximum path length" = mpl,
    "Betweenness" = b,
    "Eigenvector centrality" = e,
    "PageRank" = pr
  )

t <- rownames_to_column(t, "Airport")
t <- t %>%
  replace(is.na(.), 0) %>%
  mutate(across(where(is.numeric), round, digits = 8)) %>%
  mutate(across(where(is.numeric), as.character))
print(
  xtable(t),
  type = "latex",
  file = file.path("results", "airports_descriptors_latex_r.txt"),
  include.rownames = FALSE
)
